docs/architecture-decisions/2024-05-22_teal-compilation.md

```diff
@@ -0,0 +1,84 @@
+# Compilation of Algorand TEAL Programs into AVM Bytecode
+
+- **Status**: Proposed
+- **Owner**: Daniel McGregor (MakerX)
+- **Deciders**: Rob Moore (MakerX), Adam Chidlow (MakerX), Alessandro Cappellato (Algorand Foundation)
+- **Date created**: 2024-05-22
+- **Date decided**: 
+- **Date updated**: 
+
+## Context
+
+The Puya compiler needs to assemble valid Algorand TEAL (Transaction Execution Approval Language) programs into AVM (Algorand Virtual Machine) bytecode 
+so the bytecode can be used within inner transactions that need to create applications.
+
+There are several potential solutions to achieve this compilation, each with its own advantages and trade-offs.
+
+## Requirements
+
+* The solution works cross-platform (i.e. works on Windows, Mac and Linux)
+* The solution is fast
+* The solution is correct
+
+## Options
+
+### Option 1. Use Algod Compile Endpoint
+
+Utilize the Algorand algod `/v2/teal/compile` endpoint to compile TEAL programs into AVM bytecode.
+
+**Pros**
+
+* Ensures compatibility with the latest Algorand protocol updates.
+* Existing solution
+
+**Cons**
+
+* Requires network connectivity to the Algorand API
+* Potential latency due to network requests.
+
+### Option 2. Use Goal CLI to Compile
+
+Utilize the Algorand goal command-line interface to compile TEAL programs locally.
+
+**Pros**
+
+* Local compilation without the need for network connectivity.
+* Directly supported by Algorand, ensuring compatibility.
+
+**Cons**
+
+* Requires installation of the goal CLI
+* No native windows binary of goal is available
+
+
+### Option 3: Integrate TEAL to AVM Bytecode Assembly into Puya Compiler
+
+Extend the existing Puya compiler to directly convert TEAL programs into AVM bytecode.
+
+**Pros**
+
+* Full Control Over Compilation Process: By integrating the assembly within the Puya compiler, we maintain full control and can tailor the process to fit specific use cases and workflows.
+* No External Dependencies: This approach eliminates reliance on external tools or services, ensuring the compilation process is self-contained and not affected by network issues or external updates.
+* Additional optimizations: We can introduce additional optimizations, if desired these could be upstreamed to the algod implementation
+* Simpler Implementation: Given that Puya already handles well-structured TEAL, extending it to assemble bytecode is relatively straightforward and does not require complex parsing.
+
+**Cons**
+
+* Potential for Bugs and Inconsistencies: There is a risk of introducing bugs or deviations from Algorand's official compiler, which could lead to compatibility issues.
```

Comment:
But we can offset this risk by introducing automated tests to compare byte code matches (which we have an implementation for already)